ProxySQL 2.0.14 crashes after 10 missed heartbeats

Hi 

We are using ProxySQL version 2.0.14-70-g7b02b19 on Centos 6, (2 core and 8 GB ram) with MySQL 5.7.22 and we see very frequent crashes of ProxySQL during our peak hours as below:

2021-03-11 10:47:43 MySQL_Thread.cpp:4385:process_all_sessions(): [WARNING] Closing unhealthy client connection 
2021-03-11 10:47:43 MySQL_Thread.cpp:4385:process_all_sessions(): [WARNING] Closing unhealthy client connection 
2021-03-11 10:47:43 MySQL_Thread.cpp:4385:process_all_sessions(): [WARNING] Closing unhealthy client connection 
2021-03-11 10:47:45 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:47:48 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:47:51 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:47:54 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:47:57 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:48:00 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:48:03 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:48:06 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:48:09 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:48:12 main.cpp:1865:main(): [ERROR] Watchdog: 1 threads missed a heartbeat
2021-03-11 10:48:12 main.cpp:1869:main(): [ERROR] Watchdog: reached 10 missed heartbeats. Aborting!
2021-03-11 10:48:12 main.cpp:1870:main(): [ERROR] Watchdog: see details at https://github.com/sysown/proxysql/wiki/Watchdog
proxysql: main.cpp:1871: int main(int, const char**): Assertion `0' failed.
Error: signal 6:
/usr/bin/proxysql(_Z13crash_handleri+0x1a)[0x4d5fda]
/lib64/libc.so.6(+0x33b90)[0x7f6643ac1b90]
/lib64/libc.so.6(gsignal+0x110)[0x7f6643ac1b20]
/lib64/libc.so.6(abort+0x148)[0x7f6643ac2fc8]
/lib64/libc.so.6(+0x2c92a)[0x7f6643aba92a]
/lib64/libc.so.6(+0x2c9a2)[0x7f6643aba9a2]
/usr/bin/proxysql(main+0xc23)[0x4cd663]
/lib64/libc.so.6(__libc_start_main+0xea)[0x7f6643aaf06a]
/usr/bin/proxysql[0x4cedbe]
2021-03-11 10:48:24 main.cpp:1573:ProxySQL_daemonize_phase3(): [ERROR] ProxySQL crashed. Restarting!
2021-03-11 10:48:24 [INFO] ProxySQL version 2.0.14-70-g7b02b19

We tried loading the core dump file generated after the crash but these core dump files are truncated when we load them with gdb. We tried checking the slow query logs to see if any query is causing this issue but no such query seems to be the issue.

Please let us know if anybody from the ProxySQL team can help us with these frequent crashes.

--- a/lib/MySQL_Session.cpp
+++ b/lib/MySQL_Session.cpp
@@ -4066,8 +4066,15 @@ handler_again:
 									if (myds->myconn->MyRS && myds->myconn->MyRS->transfer_started) {
 									// transfer to frontend has started, we cannot retry
 									} else {
-										retry_conn=true;
-										proxy_warning("Retrying query.\n");
+										if (myds->myconn->mysql->server_status & SERVER_MORE_RESULTS_EXIST) {
+											// transfer to frontend has started, because this is, at least,
+											// the second resultset coming from the server
+											// we cannot retry
+											proxy_warning("Disabling query retry because SERVER_MORE_RESULTS_EXIST is set\n");
+										} else {
+											retry_conn=true;
+											proxy_warning("Retrying query.\n");
+										}
 									}
 								}
 							}

--- a/lib/mysql_connection.cpp
+++ b/lib/mysql_connection.cpp
@@ -1404,14 +1404,19 @@ handler_again:
 				} else {
 					compute_unknown_transaction_status();
 				}
+				if (_myerrno < 2000) {
+					// we can continue only if the error is coming from the backend.
+					// (or if zero)
+					// if the error comes from the client library, something terribly
+					// wrong happened and we cannot continue
+					if (mysql->server_status & SERVER_MORE_RESULTS_EXIST) {
+						async_state_machine=ASYNC_NEXT_RESULT_START;
+					}
+				}
 			}
 			if (mysql_result) {
 				mysql_free_result(mysql_result);
 				mysql_result=NULL;
-			}
-			//if (mysql_next_result(mysql)==0) {
-			if (mysql->server_status & SERVER_MORE_RESULTS_EXIST) {
-				async_state_machine=ASYNC_NEXT_RESULT_START;
 			}
 			break;
 		case ASYNC_SET_AUTOCOMMIT_START:

